[Tools][CMake] build-webkit should also check the environment to delete the cache.

clopez opened the pull request:
#### b23d0370d31e3d5f10f382a1bb7b752b9ffc09e0
<pre>
[Tools][CMake] build-webkit should also check the environment to delete the cache.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291096">https://bugs.webkit.org/show_bug.cgi?id=291096</a>

Reviewed by Adrian Perez de Castro.

In order to decide if the CMake cache should be cleaned or not, check not only the
arguments of how the script build-webkit was invoked but also if there is any
difference on the relevant variables from the environment.

And then print only a diff of what changed instead of the whole file.

* Tools/Scripts/webkitdirs.pm:
(isCachedArgumentfileOutOfDate):
(shouldRemoveCMakeCache):
(removeCMakeCache):

Canonical link: <a href="https://commits.webkit.org/294577@main">https://commits.webkit.org/294577@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e45d866810a61892b80c93302876a2e15d5d3c2a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98914 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18551 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104040 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100959 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18846 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75296 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32427 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101918 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14314 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89324 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55657 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14105 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7297 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48883 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91604 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84042 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7370 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106409 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97544 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18960 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84261 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26386 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85522 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83760 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21993 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28415 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6087 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19733 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25964 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121160 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25784 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33882 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27358 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2480,13 +2480,28 @@ sub isCachedArgumentfileOutOfDate($@)
     }
 
     open(CONTENTS_FILE, $filename);
+    local $/;
```
Excuse my Perl, but what does this do? Let's try and not make it too cryptic, we can use `$INPUT_RECORD_SEPARATOR` as an alias (as per [the documentation](https://perldoc.perl.org/variables/$%2F)). Comments are good, too:

```suggestion
    local $INPUT_RECORD_SEPARATOR;  // Slurp whole input file at once.
```

clopez replied:
I tried that but then I get an error because `$INPUT_RECORD_SEPARATOR` is undefined. It seems that `$/` is the Perl syntax to reference this and `$INPUT_RECORD_SEPARATOR` is an alias to it that only gets defined when the `English` module is imported, but we are not importing it and I would rather don't do that just for this. So I would add a comment. Refs: https://perlmaven.com/chomp

aperezdc commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2480,13 +2480,28 @@ sub isCachedArgumentfileOutOfDate($@)
+                print "+ $new_line\n";
+            }
+            $i++;
+        }
```
This may break in subtle ways if the list of environment variables changes, wouldn't it be better to have the old file lines in a set, the same for the new file lines, and then if the sets are different, subtract/intersect them to show the differences? Then the order and amount of lines is irrelevant and the logic will always work correctly. Also, using sets would work when both files contain the same lines in different order.

clopez replied:
mmm, is a bit of a corner case, the list of environment variables (or the order) shouldn't change unless the build-webkit script is edited. The environment variables are always wrote in the `build-webkit-options.txt` file (in order) and if the value is empty they are just wrote with an empty value like `CFLAGS=`
But, I will try with a set-based solution, I guess it doesn't hurt.

aperezdc reviewed (commented):
I like the idea but I think the logic to compare the files and show the differences could be more robust. Please check my comment below.

clopez committed:
[Tools][CMake] build-webkit should also check the environment to delete the cache.
https://bugs.webkit.org/show_bug.cgi?id=291096

Reviewed by Adrian Perez de Castro.

In order to decide if the CMake cache should be cleaned or not, check not only the
arguments of how the script build-webkit was invoked but also if there is any
difference on the relevant variables from the environment.

And then print only a diff of what changed instead of the whole file.

* Tools/Scripts/webkitdirs.pm:
(isCachedArgumentfileOutOfDate):
(shouldRemoveCMakeCache):
(removeCMakeCache):

Canonical link: https://commits.webkit.org/294577@main
